tools/h5part_reader.py

```diff
@@ -0,0 +1,134 @@
+
+import h5py
+import numpy as np
+
+from pydrex import minerals as _minerals
+
+
+def _get_args() -> argparse.Namespace:
+    description, epilog = __doc__.split(os.linesep + os.linesep, 1)
+    parser = argparse.ArgumentParser(description=description, epilog=epilog)
+    parser.add_argument(
+        "input",
+        help="input file (.h5part)",
+    )
+    parser.add_argument(
+        "-p",
+        "--phase",
+        help="mineral phase, default: olivine",
+        default="olivine",
+    )
+    parser.add_argument(
+        "-f",
+        "--fabric",
+        help="olivine fabric type, default: A, use 'N' for enstatite",
+        default="A",
+    )
+    parser.add_argument(
+        "-o",
+        "--output",
+        help="filename for the output NPZ file",
+    )
+    parser.add_argument(
+        "-n",
+        "--ngrains",
+        type=int,
+        help="number of grains in the polycrystal",
+        required=True,
+    )
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    args = _get_args()
+
+    if not args.input.endswith(".h5part"):
+        raise ValueError(
+            f"expected input file with .h5part extension, not {args.input}"
+        )
+    outfile = args.input[:-6] + "npz"  # Replace `.h5part` with `.npz`.
+    if type(args.output) == str and args.output != "":
+        if args.output.endswith(".npz"):
+            outfile = args.output
+        else:
+            raise ValueError(f"can only save to NPZ format, not {args.output}")
+
+    with h5py.File(args.input) as file:
+        # Don't use the last timestep,
+        # fluidity deletes the detector and just writes empty data there.
+        n_timesteps = len(file.keys()) - 1
+
+        for particle_id in file["Step#0/id"][:]:
+            option_map = {
+                "olivine": _minerals.MineralPhase.olivine,
+                "enstatite": _minerals.MineralPhase.enstatite,
+                "A": _minerals.OlivineFabric.A,
+                "B": _minerals.OlivineFabric.B,
+                "C": _minerals.OlivineFabric.C,
+                "D": _minerals.OlivineFabric.D,
+                "E": _minerals.OlivineFabric.E,
+                "N": _minerals.EnstatiteFabric.A,
+            }
+
+            # Temporary data arrays.
+            x = np.zeros(n_timesteps)
+            y = np.zeros(n_timesteps)
+            z = np.zeros(n_timesteps)
+            orientations = np.empty((n_timesteps, args.ngrains, 3, 3))
+            fractions = np.empty((n_timesteps, args.ngrains))
+
+            for t in range(n_timesteps):
+
+                # Extract particle position.
+                x[t] = file[f"Step#{t}/x"][particle_id - 1]
+                y[t] = file[f"Step#{t}/y"][particle_id - 1]
+                z[t] = file[f"Step#{t}/z"][particle_id - 1]
+
+                # Extract CPO data.
+                vals = np.empty(args.ngrains * 10)
+                for n in range(len(vals)):
+                    vals[n] = file[f"Step#{t}/CPO{n+1}"][particle_id - 1]
```

Comment:
Am I understanding correctly that you have many `CPO` particle attributes?

Reply:
Yes, the CPO data is stored on particles in a scalar array with dimension ~10000 (10 * n_grains), which is written out to 10000 single-element arrays in the h5part file. I think this is why the h5part files are so large.